staging/src/k8s.io/code-generator/cmd/validation-gen/validation.go

```diff
@@ -682,24 +721,43 @@ func (g *genValidations) emitRegisterFunction(c *generator.Context, sw *generato
 		}
 		sw.Do("scheme.AddValidationFunc((*$.rootType|raw$)(nil), func(obj, oldObj interface{}, subresources ...string) $.errorList|raw$ {\n", targs)
 		sw.Do("  if len(subresources) == 0 {\n", targs)
-		sw.Do("    return $.rootType|objectvalidationfn$(obj.(*$.rootType|raw$), nil)\n", targs)
+		sw.Do("    return $.rootType|objectvalidationfn$(obj.(*$.rootType|raw$)", targs)
+		if g.hasUpdateValidations(node) {
+			sw.Do(", oldObj.(*$.rootType|raw$), oldObj != nil", targs)
+		}
+		sw.Do(", nil)\n", targs) // fldPath
 		sw.Do("  }\n", targs)
 
 		if statusType != nil {
 			sw.Do("  if len(subresources) == 1 && subresources[0] == \"status\" {\n", targs)
 			if g.hasValidations(g.discovered.typeNodes[statusType]) {
+				hasUpdateValidations := g.hasUpdateValidations(node)
 				sw.Do("    root := obj.(*$.rootType|raw$)\n", targs)
-				sw.Do("    return $.statusType|objectvalidationfn$(&root.$.statusField$, nil)\n", targs)
+
+				// Lookup corresponding old status field value for update validation. Nil represents
+				// the absence of an old value. For create requests, the old value is always nil.
+				// Not here that the status resource SHOULD only ever be updated, so the
+				// nil check is not strictly needed.
+				if hasUpdateValidations {
+					sw.Do("    var oldStatus *$.statusType|raw$ = nil\n", targs)
```

Comment:
It could go in k8s.io/util/ptr - `ptr.SafeCall(obj, func(obj *FooSpec) *Bar { return &obj.BarField })`.  Still not sure it is better.

Reply:
You're not alone.  I also considered trying to define a utility to avoid the assignments here.  I never fully resolved it in my head, but got stuck on the "I need a utility somewhere" problem. OK if we follow up on this one?

Reply:
yes